### PR TITLE
Introduce an [AbstractSet] type.

### PR DESCRIPTION
**Pull request by bgogul**
The `AbstractSet` datastructure allows us to treat a set of values of any kind to be an `AbstractValue` in a lattice. This will be useful for data flow Analysis in general.

We are using the Kotlin sets as the underlying representation. For dataflow analysis, it would generally be more efficient to use a [persistent data structure](https://en.wikipedia.org/wiki/Persistent_data_structure) like Patricia Tries for the underling representation. For the time being, Kotlin sets should do.

### CODE REVIEW

File: javatests/arcs/core/analysis/AbstractSetTest.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    @Test
+    fun prettyPrint() {
+        assertThat("$top").isEqualTo("TOP")
+        assertThat("$bottom").isEqualTo("BOTTOM")
+        assertThat("$odds").isEqualTo("[1, 3, 5, 7, 9]")
+    }
+
+    @Test
+    fun isEquivalentTo() {
+        with(bottom) {
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/AbstractSetTest.kt:50]: Can you split this test up into three separate tests? `isEquivalentTo_bottom`, `isEquivalentTo_top`, `isEquivalentTo_odds`?

File: javatests/arcs/core/analysis/AbstractSetTest.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            assertThat(isEquivalentTo(bottom)).isFalse()
+            assertThat(isEquivalentTo(top)).isFalse()
+            assertThat(isEquivalentTo(odds)).isTrue()
+            assertThat(isEquivalentTo(evens)).isFalse()
+        }
+    }
+
+    @Test
+    fun meet() {
+        // bottom |`| something
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/AbstractSetTest.kt:72]: This test too please: `meet_bottom`, `meet_top`, `meet_odds`

File: javatests/arcs/core/analysis/AbstractSetTest.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        assertThat(odds meet bottom).isEqualTo(bottom)
+        assertThat(odds meet top).isEqualTo(odds)
+        assertThat(odds meet odds).isEqualTo(odds)
+        assertThat(odds meet evens).isEqualTo(AbstractSet<Int>(emptySet()))
+        assertThat(odds meet multiplesOfThree).isEqualTo(AbstractSet<Int>(setOf(3, 9)))
+    }
+
+    @Test
+    fun join() {
+        // bottom |_| something
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/AbstractSetTest.kt:92]: This test too

File: javatests/arcs/core/analysis/AbstractSetTest.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
+package arcs.core.analysis
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+
+/** Tests for BoundedAbstractValue. */
```
Comment (csilvestrini) [inline javatests/arcs/core/analysis/AbstractSetTest.kt:9]: Class name is wrong in this comment (you can also just delete the comment, it's not really necessary)

File: javatests/arcs/core/analysis/AbstractSetTest.kt (hunk 1 above)
Comment (bgogul) [reply javatests/arcs/core/analysis/AbstractSetTest.kt:47]: Done.

**Comment by bgogul** [comment]
Thanks for the review, @csilvestrini

**Pull request merged**